python/mlc_llm/bench/executor.py

```diff
@@ -147,9 +153,147 @@ def get_executor_feature_dict(self) -> Dict[str, Any]:
+        ]
+        base_sys_time = time.time()
+        os.environ["TOKENIZERS_PARALLELISM"] = "false"
+
+        loop = asyncio.get_running_loop()
+        pbar = None if self.disable_tqdm else tqdm(total=len(request_records))
+        with concurrent.futures.ProcessPoolExecutor(max_workers=self.num_processes) as pool:
+            futures = [
+                loop.run_in_executor(
+                    pool,
+                    FixTimestampExecutor._process_task,
+                    self.f_create_api_endpoint,
+                    partition,
+                    base_timestamp,
+                    base_sys_time,
+                    FixTimestampExecutor.launch_start_gap_time,
+                )
+                for partition in partitions
+            ]
+            results: List[RequestRecord] = []
+            for i, future in enumerate(asyncio.as_completed(futures)):
+                results.extend(await future)
+                if pbar is not None:
+                    pbar.update(len(partitions[i]))
+
+        return results
+
+    @staticmethod
+    def _process_task(
+        f_create_api_endpoint: Callable[[], APIEndPoint],
+        request_records: List[RequestRecord],
+        base_timestamp: float,
+        base_sys_time: float,
+        launch_start_gap_time: float,
+    ) -> List[RequestRecord]:
+        if len(request_records) == 0:
+            return []
+
+        async def process_task_impl(
+            f_create_api_endpoint: Callable[[], APIEndPoint],
+            request_records: List[RequestRecord],
+            base_timestamp: float,
+            base_sys_time: float,
+            launch_start_gap_time: float,
+        ) -> List[RequestRecord]:
+            api_endpoint = f_create_api_endpoint()
+            loop = asyncio.get_running_loop()
+            # Get the delta time to convert system time to the loop time.
+            # We must use the system time `time.time()` which is consistent across processes.
+            loop_sys_delta_time = loop.time() - time.time()
+            updated_request_records: List[RequestRecord] = []
+            async with api_endpoint:
+
+                async def _task(request_record: RequestRecord) -> None:
+                    updated_request_records.append(await api_endpoint(request_record))
+
+                tasks = []
+                for request_record in request_records:
+                    launch_time = (
+                        (request_record.timestamp - base_timestamp)
+                        + (base_sys_time + launch_start_gap_time)
+                        + loop_sys_delta_time
+                    )
+                    loop.call_at(
+                        launch_time,
+                        lambda record: tasks.append(asyncio.create_task(_task(record))),
+                        request_record,
+                    )
+                    # Sleep to allow runs of other scheduled tasks if any.
+                    await asyncio.sleep(0)
```

Comment:
Consider still use deferred launch with max_schedule_gap, this would help reduce burst launch in the beginning and and also reduce amount of context switch (sleep)